### ISSUE TITLE
main branch cannot build mtl/ofi on RHEL7 like

### ISSUE THREAD

**Issue by ggouaillardet** [open-mpi/ompi#10954]
`configure` passes but `make` fails:

how should this be resolved?
 - simply bump the OFI version requirements?
 - test the existence of these macros and do not build `mtl/ofi` if they are not defined?
 - test the existence of these macros and `#ifdef` protect some parts so `mtl/ofi` can be built?

FWIW
```
$ cat /etc/redhat-release 
CentOS Linux release 7.9.2009 (Core)
$ rpm -q libfabric
libfabric-1.7.2-1.el7.x86_64
```

```
In file included from /home/gilles/src/ompi-master/ompi/mca/mtl/ofi/mtl_ofi.c:13:0:
/home/gilles/src/ompi-master/ompi/mca/mtl/ofi/mtl_ofi.h: In function 'ompi_mtl_ofi_register_buffer':
/home/gilles/src/ompi-master/ompi/mca/mtl/ofi/mtl_ofi.h:324:17: error: 'struct fi_mr_attr' has no member named 'iface'
             attr.iface = FI_HMEM_CUDA;
                 ^
/home/gilles/src/ompi-master/ompi/mca/mtl/ofi/mtl_ofi.h:324:26: error: 'FI_HMEM_CUDA' undeclared (first use in this function)
             attr.iface = FI_HMEM_CUDA;
                          ^
/home/gilles/src/ompi-master/ompi/mca/mtl/ofi/mtl_ofi.h:324:26: note: each undeclared identifier is reported only once for each function it appears in
/home/gilles/src/ompi-master/ompi/mca/mtl/ofi/mtl_ofi.h:325:46: error: 'struct fi_mr_attr' has no member named 'device'
             opal_accelerator.get_device(&attr.device.cuda);
                                              ^
/home/gilles/src/ompi-master/ompi/mca/mtl/ofi/mtl_ofi.h:327:17: error: 'struct fi_mr_attr' has no member named 'iface'
             attr.iface = FI_HMEM_ROCR;
                 ^
/home/gilles/src/ompi-master/ompi/mca/mtl/ofi/mtl_ofi.h:327:26: error: 'FI_HMEM_ROCR' undeclared (first use in this function)
             attr.iface = FI_HMEM_ROCR;
                          ^
/home/gilles/src/ompi-master/ompi/mca/mtl/ofi/mtl_ofi.h:328:46: error: 'struct fi_mr_attr' has no member named 'device'
             opal_accelerator.get_device(&attr.device.cuda);
```

**Comment by bwbarrett** [comment]
Previously, we had a check for HMEM and would `#ifdef` those parts of code.  It looks like we broke that.  The HMEM requirements are pretty new in Libfabric world; I don't think we should bump up the required version of Libfabric; the old version of Libfabric should still just work with selective `#ifdef`s.

**Comment by ggouaillardet** [comment]
The `configure` absolute minimum requirement for OFI is `1.5`.

But on `v5.0.x`, the requirement is bumped to `1.9` when CUDA is detected.
There is no such thing in `main`.

`mtl_ofi_component.c` seems to requires `1.9` at runtime, but there is no `configure` check.

Should we just simply require `1.9` at `configure` time?

**Comment by jsquyres** [comment]
@ggouaillardet Has this been fixed yet?  It would seem to be a problem if we can't compile out-of-the-box with the RHEL 7 libfabric.

**Comment by janciesko** [comment]
Seeing the same issue on our cluster. (Red Hat Enterprise Linux Server release 7.6 (Maipo))

**Comment by jsquyres** [comment]
FYI @open-mpi/ofi team.  Are you comfortable requiring libfabric >= v1.9, especially given that that will disqualify RHEL7 inbox libfabric?

**Comment by hppritcha** [comment]
Not really @jsquyres .  I'd prefer an ifdef approach too.

**Comment by jsquyres** [comment]
It would be great if someone in @open-mpi/ofi would fix this, then. 😇

**Comment by hppritcha** [comment]
yes i'm checking this out on an old rhel7+libfab 1.7 system.

**Comment by wckzhang** [comment]
There is fallback code for the getinfo, we can change the minimum required version when FI_HMEM hints are not detected

**Comment by hppritcha** [comment]
Do we want to require libfabric 1.9 or do we want to be able to use the libfabric that ships with RHEL7?

**Comment by wckzhang** [comment]
I'm okay with requiring libfabric 1.9

**Comment by janciesko** [comment]
4110432d436c62cb6aafdc8b04c3c25b0749fe3a fixes the issue for me.

**Comment by s417-lama** [comment]
I think v1.9 is not enough, because `FI_HMEM_ROCR` was added from v1.11.0 (https://github.com/ofiwg/libfabric/commit/3aff7fc138c56c9d13a72e9ac4eccc69de686c48).

I got the following error:
```
Making install in mca/mtl/ofi
make[2]: Entering directory '/sqfs2/cmc/1/work/G15442/v60680/.kochi/projects/ityrbench/install_src/squid-c/openmpi/main/ityrbench/build/ompi/mca/mtl/ofi'
  CC       mtl_ofi.lo
In file included from ../../../../../ompi/mca/mtl/ofi/mtl_ofi.c:13:
../../../../../ompi/mca/mtl/ofi/mtl_ofi.h: In function ‘ompi_mtl_ofi_register_buffer’:
../../../../../ompi/mca/mtl/ofi/mtl_ofi.h:328:26: error: ‘FI_HMEM_ROCR’ undeclared (first use in this function); did you mean ‘FI_HMEM_CUDA’?
             attr.iface = FI_HMEM_ROCR;
                          ^~~~~~~~~~~~
                          FI_HMEM_CUDA
../../../../../ompi/mca/mtl/ofi/mtl_ofi.h:328:26: note: each undeclared identifier is reported only once for each function it appears in
make[2]: *** [Makefile:1590: mtl_ofi.lo] Error 1
make[2]: Leaving directory '/sqfs2/cmc/1/work/G15442/v60680/.kochi/projects/ityrbench/install_src/squid-c/openmpi/main/ityrbench/build/ompi/mca/mtl/ofi'
make[1]: *** [Makefile:2792: install-recursive] Error 1
make[1]: Leaving directory '/sqfs2/cmc/1/work/G15442/v60680/.kochi/projects/ityrbench/install_src/squid-c/openmpi/main/ityrbench/build/ompi'
make: *** [Makefile:1524: install-recursive] Error 1
```

My libfabric version is:
```console
> rpm -q libfabric
libfabric-1.9.0rc1-1.el8.x86_64
```

**Comment by lrbison** [comment]
I agree with @bwbarrett and @jsquyres.  We will take a look and see if we can replace the `#ifdef`'s again

**Comment by jsquyres** [comment]
> I agree with @bwbarrett and @jsquyres. We will take a look and see if we can replace the `#ifdef`'s again

Thanks @lrbison!  If we can restore/fix those `#if`/`#ifdef`s, can we re-evaluate whether restricting Libfabric to >= v1.9 (via 4110432d436c62cb6aafdc8b04c3c25b0749fe3a) is really necessary?

**Comment by hppritcha** [comment]
My thinking is we should not have too many ifdefs.  The more we have, the more likely we won't be testing all "supported" cases.

**Comment by jsquyres** [comment]
Fair enough. However:

1. I don't think `#if`'s are inherently evil.
2. We probably don't want to automatically disqualify the inbox libfabric on popular distros (e.g., RHEL 7, which -- as reported earlier in this issue -- has Libfabric 1.7).

Because of 2, we might *have* to use the `#if` approach.

But it should probably also be reconciled with other statements made in this issue: 

* But on v5.0.x, the requirement is bumped to 1.9 when CUDA is detected.  There is no such thing in `main`.
* mtl_ofi_component.c seems to requires 1.9 at runtime, but there is no `configure` check.
* I think v1.9 is not enough, because `FI_HMEM_ROCR` was added from v1.11.0 (https://github.com/ofiwg/libfabric/commit/3aff7fc138c56c9d13a72e9ac4eccc69de686c48).

Meaning: both the `FI_HMEM_*` stuff should probably be protected with appropriate `#if`s, and how `mtl_ofi_component.c` determines what API level to ask for should probably be audited to ensure it's doing what people think it should be doing.

**Comment by hppritcha** [comment]
The second bullet above is addressed by #11111
